src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/FlowFetcher.java

```diff
@@ -122,44 +123,83 @@ protected void handleInput(Tuple input) throws Exception {
     }
 
     private void updatePeriodicPingHeap(Tuple input) throws PipelineException {
+        if (hasHaFlowPeriodicPingRequest(input)) {
+            updatePeriodicHaPingHeap(input);
+            return;
+        }
+
         PeriodicPingCommand periodicPingCommand = pullPeriodicPingRequest(input);
+
         if (periodicPingCommand.isEnable()) {
             flowRepository.findById(periodicPingCommand.getFlowId())
                     .flatMap(flow -> {
                         flowRepository.detach(flow);
                         return getFlowWithTransitEncapsulation(flow);
                     })
                     .ifPresent(flowsSet::add);
+
         } else {
             flowsSet.removeIf(flowWithTransitEncapsulation ->
                     flowWithTransitEncapsulation.getFlow().getFlowId().equals(periodicPingCommand.getFlowId()));
         }
     }
 
+    private void updatePeriodicHaPingHeap(Tuple input) throws PipelineException {
+        PeriodicHaPingCommand periodicHaPingCommand = pullPeriodicHaPingRequest(input);
+        String haFlowId = periodicHaPingCommand.getHaFlowId();
+        if (periodicHaPingCommand.isEnable()) {
+            haFlowRepository.findById(haFlowId)
+                    .flatMap(haFlow -> {
+                        haFlowRepository.detach(haFlow);
+                        return getFlowWithTransitEncapsulation(haFlow);
+                    })
+                    .ifPresent(flowsSet::add);
+
+        } else {
+            flowsSet.removeIf(flowWithTransitEncapsulation ->
+                    flowWithTransitEncapsulation.getHaFlow().getHaFlowId().equals(haFlowId));
+        }
+    }
+
+
     private void refreshHeap(Tuple input, boolean emitCacheExpiry) throws PipelineException {
-        log.debug("Handle periodic ping request");
-        Set<FlowWithTransitEncapsulation> flowsWithTransitEncapsulation =
-                flowRepository.findWithPeriodicPingsEnabled().stream()
-                        .peek(flowRepository::detach)
-                        .map(this::getFlowWithTransitEncapsulation)
-                        .flatMap(o -> o.isPresent() ? Stream.of(o.get()) : Stream.empty())
-                        .collect(Collectors.toSet());
+        Set<FlowWithTransitEncapsulation> flowsWithTransitEncapsulation = getFlowsWithTransitEncapsulation();
+        Set<FlowWithTransitEncapsulation> haFlowsWithTransitEncapsulation = getHaFlowsWithTransitEncapsulation();
         if (emitCacheExpiry) {
             final CommandContext commandContext = pullContext(input);
             emitCacheExpire(input, commandContext, flowsWithTransitEncapsulation);
         }
         flowsSet = flowsWithTransitEncapsulation;
```

Comment:
`flowsSet` is declared in line 92. This Set is used to keep a cache of the flows which have to be pinged periodically. It is refreshed automatically every `periodicPingCacheExpiryInterval` seconds which by default is 60s. 

In case of a creation of a new flow or HaFlow using each creation service, it can be manually added to the flowSet. This is done by the `updatePeriodicPingHeap` method.

Reply:
I guess then we can keep the current behaviour. The set is initialised for the first time at deployment time.